### PR TITLE
Remove useless reshape

### LINKED ISSUE #2752: Opt: Reshape(arange(x)) -> arange(x) when this do not change the shapes, number of dimensions.
```
>>> import theano
>>> i=theano.tensor.iscalar()
>>> m=theano.tensor.mgrid[0:i,]
>>> f=theano.function([i], m)
>>> theano.printing.debugprint(f, print_type=True)
Reshape{1} [@A] <TensorType(int32, vector)> ''   3
 |ARange [@B] <TensorType(int32, vector)> ''   1
 | |TensorConstant{0} [@C] <TensorType(int8, scalar)>
 | |<TensorType(int32, scalar)> [@D] <TensorType(int32, scalar)>
 | |TensorConstant{1} [@E] <TensorType(int8, scalar)>
 |MakeVector [@F] <TensorType(int64, vector)> ''   2
   |Elemwise{Cast{int64}} [@G] <TensorType(int64, scalar)> ''   0
     |<TensorType(int32, scalar)> [@D] <TensorType(int32, scalar)>
```

This happen in gh-2572

Make this more general, in the sense that if the input is 1d and the output is 1d, what ever created it, remove it (add AssertOp if we can't assert that the shape are the same)

### PR DESCRIPTION
**Pull request by yingzha**
Fixes #2752

### CODE REVIEW

File: theano/tensor/tests/test_opt.py [hunk 1]
```diff
@@ -5719,6 +5719,16 @@ def test_local_reshape(self):
         assert sum(isinstance(node.op, self.op) for node in topo) == 1
 
 
+def test_local_useless_reshape():
+    mode = theano.compile.get_default_mode().including(
+            'local_useless_reshape')
+    i = T.iscalar('i')
+    m=theano.tensor.mgrid[0:i,]
```
Comment (abergeron) [inline theano/tensor/tests/test_opt.py]: Small nit: there should be spaces around the `=`

Comment (yingzha) [inline theano/tensor/tests/test_opt.py]: Oh, I forgot it.

**Comment by abergeron** [comment]
You should probably check the broadcastable pattern of the input and output and leave the reshape if it's not the same.  This causes problems in some cases otherwise.

**Comment by nouiz** [comment]
I think if the broadcast pattern change, then we need to introduce a Rebroadcast op. This can be done easily by using the method patternbroadcast()

**Comment by yingzha** [comment]
@abergeron , I just added a judgement to compare the broadcastable pattern, I'm not sure if it's wise nor not to do so.

**Comment by abergeron** [comment]
Code is fine, all that is left to do is to remove the merge commits.

**Comment by yingzha** [comment]
@abergeron done.

**Comment by abergeron** [comment]
Ok to merge when travis passes.

**Pull request merged**